### PR TITLE
[BC-BREAKING] Add align_corners option to grid_sample and affine_grid, change default to False

### LINKED ISSUE #20785: grid_sample is not aligned
`F.interpolate` has been changed to use `align_corners=False` by default, which is also the pixel model used by opencv/PIL, etc.

The input to `grid_sample` is in range `[-1, 1]`, and therefore should ideally be scale-invariant.
However, `F.grid_sample` still uses the old pixel model: https://github.com/pytorch/pytorch/blob/65b00aa5972e23b2a70aa60dec5125671a3d7153/aten/src/ATen/native/cuda/GridSampler.cu#L161-L162, and as a result it is not properly aligned under the scaling of `F.interpolate(align_corners=False)`. 

Examples:

```python
import torch
from torch.nn import functional as F

def grid_sample_v2(input, grid):
    # grid: [-1, 1]
    N, C, H, W = input.shape
    gridx = grid[:, :, :, 0]
    gridy = grid[:, :, :, 1]
    gridx = ((gridx + 1) / 2 * W - 0.5) / (W - 1) * 2 - 1
    gridy = ((gridy + 1) / 2 * H - 0.5) / (H - 1) * 2 - 1
    newgrid = torch.stack([gridx, gridy], dim=-1)
    return F.grid_sample(input, newgrid)

N = 500
input = torch.rand(N, 1, 32, 74)
input2 = F.interpolate(input, scale_factor=5, align_corners=False, mode='bilinear')

grid = torch.rand(N, 18, 17, 2) * 1.8 - 0.9  # use coordinates in -0.9, 0.9 to avoid boundary effects

output = F.grid_sample(input, grid)
output2 = F.grid_sample(input2, grid)
diff = torch.abs(output2 - output)
print("DIFFv1:", diff.mean())

output = grid_sample_v2(input, grid)
output2 = grid_sample_v2(input2, grid)
diff = torch.abs(output2 - output)
print("DIFFv2:", diff.mean())
```
The above code, when run with `align_corners=True`, prints:
```
DIFFv1: tensor(0.0032)
DIFFv2: tensor(0.0675)
```
when run with `align_corners=False`, prints:
```
DIFFv1: tensor(0.0689)
DIFFv2: tensor(5.3741e-07)
```
In the example, I provide a new implementation of `grid_sample`, which leads to smaller reconstruction error when `align_corners=False`. Given that `align_corners=False` is now the default for `interpolate`, similar options should be built into `grid_sample` as well.

### LINKED ISSUE #24821: Document affine_grid works with 3d
See https://github.com/pytorch/pytorch/issues/24380#issuecomment-522294803

### PR DESCRIPTION
**Pull request by bnehoran**
Resolves: https://github.com/pytorch/pytorch/issues/20785
Addresses https://github.com/pytorch/pytorch/issues/24470 for `affine_grid`
Subsumes and closes: https://github.com/pytorch/pytorch/pull/24878 and likewise closes: https://github.com/pytorch/pytorch/issues/24821

Adds the `align_corners` option to `grid_sample` and `affine_grid`, paralleling the option that was added to `interpolate` in version 0.4.0.

In short, setting `align_corners` to `False` allows these functions to be resolution agnostic.
This ensures, for example, that a grid generated from a neural net trained to warp 1024x1024 images will also work to warp the same image upsampled/downsampled to other resolutions like 512x512 or 2048x2048 without producing scaling/stretching artifacts.
 
Refer to the documentation and #20785 for more details.

#### BC-Breaking Changes

- **Important**: BC-Breaking change because of new default for `align_corners`
The old functionality can still be achieved by setting `align_corners=True`, but the default is now set to `align_corners=False`, since this is the more correct setting, and since this matches the default setting of `interpolate`.

- **Should not cause BC issues**: BC-Breaking change for pathological use case
2D affine transforms on 1D coordinates and 3D affine transforms on 2D coordinates (that is, when one of the spatial dimensions has an empty span) are ill-defined, and not an intended use case of `affine_grid`. Whereas before, all grid point components along such dimension were set arbitrarily to `-1` (that is, before multiplying be the affine matrix), they are now all set instead to `0`, which is a much more consistent and defensible arbitrary choice. A warning is triggered for such cases.

#### Documentation

- Update `affine_grid` documentation to express that it does indeed support 3D affine transforms. This support was already there but not documented.
- Add documentation warnings for BC-breaking changes in `grid_sample` and `affine_grid` (see above).

#### Refactors

- `affine_grid` no longer dispatches to cuDNN under any circumstances.
The decision point for when the cuDNN `affine_grid_generator` is compatible with the native PyTorch version and when it fails is a headache to maintain (see [these conditions](https://github.com/seung-lab/pytorch/blob/5377478e94a7e2c86181a7686a7a492ecd147815/torch/nn/_functions/vision.py#L7-L8)). The native PyTorch kernel is now used in all cases.

- The kernels for `grid_sample` are slightly refactored to make maintenance easier.

#### Tests
Two new tests are added in `test_nn.py`:
- `test_affine_grid_error_checking` for errors and warnings in `affine_grid`
- `test_affine_grid_3D` for testing `affine_grid`'s 3D functionality. The functionality existed prior to this, but wasn't tested.

### CODE REVIEW

**Comment by bnehoran** [comment]
This is a resubmit of https://github.com/pytorch/pytorch/pull/23923 

cc @SsnL @ailzhang

File: test/test_nn.py [hunk 1]
```diff
@@ -7956,52 +8071,119 @@ def test_shape(N, C, ID, IH, IW, D, H, W, mode, padding_mode):
...
+                gradients = torch.randn(out_cpu.size())
+                out_cpu.backward(gradients)
+                input_gpu = input_cpu.detach().cuda().requires_grad_()
+                with warnings.catch_warnings(record=True):
+                    out_cuda = F.affine_grid(input_gpu, sz, align_corners=align_corners)
+                out_cuda.backward(gradients.cuda())
+                self.assertEqual(out_cpu, out_cuda)
+                self.assertEqual(input_cpu.grad, input_gpu.grad)
+
+    def test_affine_grid_3D(self):
```
Comment (bnehoran) [inline test/test_nn.py]:
```suggestion
    def test_affine_grid_3d(self):
```

**Comment by bnehoran** [comment]
The failed test case in `pytorch_linux_trusty_py2_7_9_test` seems to be that the empty span warning (details in the PR description) is not getting triggered for some reason.

The corresponding python3 CI test, `pytorch_linux_trusty_py3_6_gcc5_4_test`, does pass the same test case, and triggers the warning, so it seems that the problem might be confined to python2.

When I run python2 locally, however, the warning _does_ get triggered, and so I am currently not able to reproduce the failed test case.

**Review by ailzhang — APPROVED** [review]
Thanks!

**Comment by krikru** [comment]
Shouldn't the inclusion of this fix necessitate a new major version, since it breaks backwards compatibility?

**Pull request closed without merging**